File: IntegrationTests/Parser/TypeParsing/TypeParserTests/Closure/Closure_TypeParserTest.cs
// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using Microsoft.FxCop.Sdk;
using NUnit.Framework;

namespace InjectionCop.IntegrationTests.Parser.TypeParsing.TypeParserTests.Closure
{
  [TestFixture]
  public class Closure_TypeParserTest : TypeParserTestBase
  {
    [Test]
    public void Parse_SafeClosureUsingLocalVariable_NoProblem ()
    {
      Method sample = TestHelper.GetSample<ClosureSample> ("SafeClosureUsingLocalVariable");
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.False);
    }

    [Test]
    public void Parse_UnsafeClosureUsingLocalVariable_ReturnsProblem ()
    {
      Method sample = TestHelper.GetSample<ClosureSample> ("UnsafeClosureUsingLocalVariable");
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.True);
    }

    [Test]
    public void Parse_SafeClosureUsingField_NoProblem ()
    {
      Method sample = TestHelper.GetSample<ClosureSample> ("SafeClosureUsingField");
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.False);
    }

    [Test]
    public void Parse_UnsafeClosureUsingField_ReturnsProblem ()
    {
      Method sample = TestHelper.GetSample<ClosureSample> ("UnsafeClosureUsingField");
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.True);
    }
  }
}

File: IntegrationTests.VB/Parser/TypeParsing/TypeParserTests/Closure/Closure_TypeParserTest.vb
Imports Microsoft.FxCop.Sdk
Imports NUnit.Framework
Imports System

Namespace Parser.TypeParsing.TypeParserTests.Closure
	<Ignore("Closures are not suppoted in version 1.0"), TestFixture()>
	Public Class Closure_TypeParserTest
		Inherits TypeParserTestBase

		<Test()>
		Public Sub Parse_SafeClosureUsingLocalVariable_NoProblem()
			Dim sample As Method = TestHelper.GetSample(Of ClosureSample)("SafeClosureUsingLocalVariable", New TypeNode() {})
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[False])
		End Sub

		<Test()>
		Public Sub Parse_UnsafeClosureUsingLocalVariable_ReturnsProblem()
			Dim sample As Method = TestHelper.GetSample(Of ClosureSample)("UnsafeClosureUsingLocalVariable", New TypeNode() {})
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[True])
		End Sub

		<Test()>
		Public Sub Parse_SafeClosureUsingField_NoProblem()
			Dim sample As Method = TestHelper.GetSample(Of ClosureSample)("SafeClosureUsingField", New TypeNode() {})
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[False])
		End Sub

		<Test()>
		Public Sub Parse_UnsafeClosureUsingField_ReturnsProblem()
			Dim sample As Method = TestHelper.GetSample(Of ClosureSample)("UnsafeClosureUsingField", New TypeNode() {})
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[True])
		End Sub
	End Class
End Namespace
